boskos/ranch/priority.go

```diff
@@ -0,0 +1,219 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ranch
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// request stores request information with expiration
+type request struct {
+	id         string
+	expiration time.Time
+}
+
+func newRequestQueue() *requestQueue {
+	return &requestQueue{
+		requestMap: map[string]request{},
+	}
+}
+
+// requestQueue is a simple FIFO queue for requests.
+type requestQueue struct {
+	lock        sync.RWMutex
+	requestList []string
+	requestMap  map[string]request
+}
+
+// update updates expiration time is updated if already present,
+// add a new requestID at the end otherwise (FIFO)
+func (rq *requestQueue) update(requestID string, newExpiration time.Time) {
+	if requestID == "" {
+		// Nothing to do
+		return
+	}
+	rq.lock.Lock()
+	defer rq.lock.Unlock()
+	if rq.requestMap == nil {
+		rq.requestMap = map[string]request{}
+	}
+	req, exists := rq.requestMap[requestID]
+	if !exists {
+		req = request{id: requestID}
+		rq.requestList = append(rq.requestList, requestID)
+		logrus.Infof("request id %s added", requestID)
+	}
+	// Update timestamp
+	req.expiration = newExpiration
+	rq.requestMap[requestID] = req
+	logrus.Infof("request id %s set to expire at %v", requestID, newExpiration)
+}
+
+// delete an element
+func (rq *requestQueue) delete(requestID string) {
+	rq.lock.Lock()
+	defer rq.lock.Unlock()
+	delete(rq.requestMap, requestID)
+	index := -1
+	for i, id := range rq.requestList {
+		if id == requestID {
+			index = i
+			break
+		}
+	}
+	if index >= 0 {
+		rq.requestList = append(rq.requestList[0:index], rq.requestList[index+1:len(rq.requestList)]...)
```

Comment:
with a linked list we would not need to reslice, do we care about memory footprint here?

Reply:
good point. Fixed